feat: migrate from MySQL C API to MySQL X DevAPI

dudantas opened the pull request:
This pull request encompasses the migration from the traditional MySQL C API to the modern MySQL X DevAPI across our project. This change introduces enhanced robustness, improved code readability, and leverages the modern features of MySQL X DevAPI.

Key Changes:

• Replaced all instances of MySQL C API function calls with their MySQL X DevAPI equivalents.
• Refactored database interaction logic to utilize the fluent API provided by MySQL X DevAPI, ensuring more maintainable and clearer code.
• Implemented error handling and data type conversions using MySQL X DevAPI's robust mechanisms, enhancing stability and performance.
• Updated corresponding documentation and comments to reflect changes in the database access layer.

This update paves the way for future database interactions to be more flexible and easier to manage.

To test on Windows, you need to install the official MySQL and preferably the latest Nginx and PHP using Chocolatey.

Important: in mysqlx, the port used is 33060

dudantas committed:
perf: PreparedStatement usage in database queries

Co-Authored-By: Beats <daniel15042015@gmail.com>

kaleohanopahala commented:
Wouldn't it be better to use std::unique_ptr to avoid memory leaks? And std::lock_guard for being safer?

```
std::unique_ptr<MYSQL_STMT, decltype(&mysql_stmt_close)> stmt(mysql_stmt_init(handle), &mysql_stmt_close);
if (!stmt) {
    g_logger().error("Failed to initialize the statement handle: {}", mysql_error(handle));
    return nullptr;
}
```

```
std::unique_ptr<MYSQL_RES, decltype(&mysql_free_result)> result(mysql_stmt_result_metadata(stmt.get()), &mysql_free_result);
if (!result) {
    return nullptr;
}
```

W4gNII commented:
![image](https://github.com/opentibiabr/canary/assets/108229470/68e6f884-5b7f-4c48-a3f0-b2b4692351c6)
testing this pr after player logout if you preffer i have a full log in txt

dudantas committed:
feat: migrate from MySQL C API to MySQL X DevAPI

This pull request encompasses the migration from the traditional MySQL C API to the modern MySQL X DevAPI across our project. This change introduces enhanced robustness, improved code readability, and leverages the modern features of MySQL X DevAPI.

Key Changes:

• Replaced all instances of MySQL C API function calls with their MySQL X DevAPI equivalents.
• Refactored database interaction logic to utilize the fluent API provided by MySQL X DevAPI, ensuring more maintainable and clearer code.
• Implemented error handling and data type conversions using MySQL X DevAPI's robust mechanisms, enhancing stability and performance.
• Updated corresponding documentation and comments to reflect changes in the database access layer.

This update paves the way for future database interactions to be more flexible and easier to manage.

To test on Windows, you need to install the official MySQL and preferably the latest Nginx and PHP using Chocolatey.

dudantas commented:
> testing this pr after player logout if you preffer i have a full log in txt

I changed everything to use mysqlx, I haven't implemented the prepared statement yet, but it's working.

> Wouldn't it be better to use std::unique_ptr to avoid memory leaks? And std::lock_guard for being safer?

Same here, the quoted part of the code has been removed.

dudantas committed:
feat: add documentation and test use of "get" database functions

dudantas committed:
feat: simplify some database functions

dudantas committed:
refactor: PropStream for better encapsulation and type safety

This commit introduces several key improvements to the PropStream and DBResult classes to enhance maintainability, type safety, and performance:

1. Encapsulation of Template Functions: Moved template function implementations from header files to .cpp files. This change reduces compilation dependencies and improves compile times by minimizing recompilation needs when templates are modified.

2. Specialized Read Functions: Replaced generic template read functions with type-specific methods such as readU8, readU16, etc. This approach enhances type safety, allowing for stricter type checks and reducing runtime errors.

3. Vector Instead of Raw Pointers: Replaced raw pointer parameters with std::vector<uint8_t> in PropStream's handling of binary data. This change increases safety by leveraging C++ container management features, avoiding manual memory management and potential data corruption issues.

These changes are aimed at making the codebase more robust, easier to manage, and less prone to errors, particularly in the handling of binary data and type conversions.

dudantas committed:
fix: string view

dudantas committed:
feat: prepared statement on kv system

dudantas committed:
fix: remove duplicate check

dudantas committed:
fix: compilation error

dudantas committed:
fix: build ubuntu errors

dudantas committed:
fix: circular inclusion related to database

dudantas committed:
reorganize functions

dudantas committed:
fix: prepared statement bug

dudantas committed:
fix: gha check

dudantas committed:
fix: compilation gha

dudantas committed:
fix: tripet and missing column

dudantas committed:
fix: compilation

majestyotbr commented:
Update your pull request and use the pull request template.

luan committed:
dudantas/improve-add-database-caching

luan committed:
fix compilation

luan committed:
fix linux linking

dudantas committed:
fix: indent

dudantas committed:
fix

elsongabriel committed:
Merge branch 'main' into dudantas/improve-add-database-caching

dudantas committed:
Merge branch 'main' into dudantas/improve-add-database-caching

dudantas committed:
Merge branch 'main' into dudantas/improve-add-database-caching

dudantas committed:
fix: builtin baseline

dudantas committed:
fix: compilation

dudantas committed:
fix: compilation

dudantas committed:
fix: resultCount function, unit tests and others

dudantas committed:
fix: remove version tag

dudantas committed:
update: buildtin baseline

dudantas committed:
fix: getVIPEntries

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
fix: reorder

dudantas committed:
fix: tests

dudantas committed:
fix: test errors

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
fix: ports

dudantas committed:
Merge branch 'main' into dudantas/improve-add-database-caching

dudantas committed:
fix: some reverts

dudantas committed:
fix: database tests

dudantas committed:
fix: remove duplicated IPO

dudantas committed:
fix: database tests

dudantas committed:
Apply suggestions from code review

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

dudantas committed:
fix: dockerfile

dudantas committed:
fix: change updateStatistics to mysqlx

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
improve: abstract get functions and try/catch from mysqlx

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
fix: correct function

dudantas committed:
fix: some mysqlx devapi changes

dudantas committed:
Merge branch 'main' into dudantas/improve-add-database-caching

dudantas committed:
fix: compilation and indentation

dudantas committed:
fix: some things

dudantas committed:
Merge branch 'main' into dudantas/improve-add-database-caching

dudantas committed:
fix: some things

dudantas commented:
After reviewing the changes, I’ve concluded that fully migrating from the MySQL C API to MySQL X DevAPI is more complex than initially anticipated. While the modern features and improved design are appealing, the implementation effort and potential disruptions currently outweigh the benefits. Therefore, I’m closing this PR for now. We may revisit this approach in the future once the project’s priorities and resources align better with such a significant transition.
